### PR TITLE
Use redcap_event_name to split visit types

### PR DESCRIPTION
**Pull request by s-emerson**
This change introduces a sorting algorithm for an input .csv file into the nacculator program. This means that nacculator is able to parse which records in a .csv file should be processed depending on which flags are active (-ivp vs -fvp, -ftld, and so on), and so the splitting step in the [NACC Upload Process](https://ctsit-forge.ctsi.ufl.edu/projects/adrc/wiki/NACC_Upload_Process)
`bash tools/preprocess/split_ivp_fvp.sh run_01-11-2019/final_Update.csv`
can be skipped entirely.

This change was made to make the NACC upload process smoother and easier for users. It allows users to export all the data for an ADRC REDCap project together as one .csv so that it can be run through NACCulator without the need to create separate files for each flag.

## Verification

1. From the command-line, run `source venv/bin/activate` to activate the virtual environment.
2. Run `redcap2nacc -ivp <data.csv >data.txt` with your choice of flag and input .csv file.
3. The output should look like a .txt file with no blank lines, only the correct processed forms.
4. If you pass a .csv file that does not contain the forms and redcap_event_name to match the flag, the output should be a blank file, because all records that did not match the flag's redcap_event_name were skipped.
5. The output .txt files for each flag are ready to be uploaded to NACC as usual.

- Verify the thing does this: Creates the same output .txt files as the previous version of nacculator without the need to split the REDCap .csv file beforehand.
- Verify the thing does not do that: Print blank lines for records that do not contain the expected forms.

## Affirmations

All of these should have a check by them. Any exception requires an explanation.

* [x] I matched the style of the existing code.
* [x] I added and updated any relevant documentation (inline, `README`, `CHANGELOG`, and such).
* [x] I used Python's type hinting.
* [x] I ran the automated tests and ensured they **ALL** passed.
* [x] I ran the linter and ensured my changes have not introduced **ANY** warnings or errors.
* [x] I have made an effort to minimize code changes and have not included any cruft (preference files, *.pyc files, old comments, print-debugging, unused variables).
* [x] I have made an effort maintain a clear commit history (haven't merged other branches or rebased improperly).
* [x] I have written the code myself or have given credit where credit is due.

### CODE REVIEW

File: tests/test_check_redcap_event.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    csf = False
+    np = False
+    m = False
+    ivp = False
+    fvp = False
+
+
+class TestRedcapEvent(unittest.TestCase):
+    '''
+    These tests are meant to ensure that the test_redcap_event function is
```
Comment (ufl-taeber) [inline tests/test_check_redcap_event.py:18]: Did you mean `check_redcap_event`?
```suggestion
    These tests are meant to ensure that the check_redcap_event function is
```

File: nacc/redcap2nacc.py [hunk 2]
```diff
@@ -146,6 +146,69 @@ def check_for_bad_characters(field: Field) -> typing.List:
...
+                (form_match_z1x == '0' or form_match_z1x == ''):
+            event_match = False
+            return event_match
+    elif options.tfp:
+        event_name = 'telephone_followup'
+    elif options.m:
+        event_name = 'milestone'
+
+    redcap_event = record['redcap_event_name']
+    event_match = re.search(event_name, redcap_event)
```
Comment (ufl-taeber) [inline nacc/redcap2nacc.py:208]: Consider
```suggestion
    event_match = event_name in redcap_event
```

File: nacc/redcap2nacc.py [hunk 3]
```diff
@@ -247,6 +310,11 @@ def convert(fp, options, out=sys.stdout, err=sys.stderr):
     """Converts data in REDCap's CSV format to NACC's fixed-width format."""
     reader = csv.DictReader(fp)
     for record in reader:
+        if not options.csf:
+            event_match = check_redcap_event(options, record)
+            if not event_match:
+                continue
```
Comment (ufl-taeber) [inline nacc/redcap2nacc.py:316]: Please log a message here stating the reason why the record was skipped.

File: nacc/redcap2nacc.py [hunk 4]
```diff
@@ -146,6 +146,69 @@ def check_for_bad_characters(field: Field) -> typing.List:
...
 
+def check_redcap_event(options, record) -> bool:
+    """
+    Determines if the record's redcap_event_name and filled forms match the
+    options flag
+    """
+    if options.lbd and options.ivp:
+        event_name = 'initial_visit'
+        form_match_lbd = record['lbd_ivp_b1l_complete']
+        if (form_match_lbd == '0' or form_match_lbd == ''):
```
Comment (ufl-taeber) [inline nacc/redcap2nacc.py:157]: Please remove the unnecessary parenthesis here and elsewhere in this function
```suggestion
        if form_match_lbd == '0' or form_match_lbd == '':
```

File: nacc/redcap2nacc.py [hunk 5]
```diff
@@ -146,6 +146,69 @@ def check_for_bad_characters(field: Field) -> typing.List:
...
+            event_match = False
+            return event_match
```
Comment (ufl-taeber) [inline nacc/redcap2nacc.py:159]: Consider simplifying these returns.
```suggestion
            return False
```

File: nacc/redcap2nacc.py [hunk 6]
```diff
@@ -146,6 +146,69 @@ def check_for_bad_characters(field: Field) -> typing.List:
...
+    # Right now it is a single non-longitudinal form in a REDCap project with
+    # no redcap_event_names.
+    # elif options.csf:
+    #     event_name = 'csf_'
+    elif options.ivp:
+        event_name = 'initial_visit'
+        form_match_z1 = record['ivp_z1_complete']
+        form_match_z1x = record['ivp_z1x_complete']
+        if (form_match_z1 == '0' or form_match_z1 == '') and \
+                (form_match_z1x == '0' or form_match_z1x == ''):
```
Comment (ufl-taeber) [inline nacc/redcap2nacc.py:189]: In Python, you can test multiple conditions like this:
```suggestion
        if form_match_z1 in ['', '0'] and form_match_z1x in ['', '0']:
```

**Comment by ufl-taeber** [comment]
Do you have some test data I can use to rerun your test?

**Comment by ufl-taeber** [comment]
Are the event names the same at different sites? We might have to make them configurable.

File: tests/test_check_redcap_event.py (hunk 1 above)
Comment (s-emerson) [reply tests/test_check_redcap_event.py]: good catch, thank you.

File: nacc/redcap2nacc.py (hunk 5 above)
Comment (s-emerson) [reply nacc/redcap2nacc.py]: this makes the code look much cleaner, thanks

**Comment by ufl-taeber** [comment]
This seems to work as intended. When I ran `redcap2nacc -np <data.csv` with the test data you sent me, I got this error: `ValueError: "4" is unacceptable for NPGRCCA`. Maybe something to look into, but not a blocker for merging.

**Pull request merged**